### PR TITLE
feat: add azure-arc-servers skill

### PR DESCRIPTION
**Pull request by deankroker**
## Summary

Adds a new `azure-arc-servers` skill that routes Azure Arc–enabled server (Connected Machine) intents — onboarding, troubleshooting, and day-2 management — to the right workflow.

The skill targets servers that live **outside Azure** (on-prem, AWS, GCP, edge) being projected into Azure via the Connected Machine agent (`azcmagent`) — ARM type `Microsoft.HybridCompute/machines`. It is explicitly disambiguated from `azure-compute` (which owns brand-new VMs in Azure proper, `Microsoft.Compute/virtualMachines`).

## Structure

```
plugin/skills/azure-arc-servers/
├── SKILL.md                                      # router (intent → workflow)
├── version.json                                  # NBGV: 1.0 → stamped 1.0.0 at build
├── references/
│   ├── arc-vs-azure-vm.md                        # disambiguation cheatsheet
│   └── arc-mcp-tools.md                          # Azure MCP + az + azcmagent lookup
└── workflows/
    ├── arc-server-onboard/
    │   └── references/
    │       ├── prerequisites.md
    │       ├── connectivity-options.md           # Public / Private Link / Arc Gateway / Edge / air-gapped
    │       ├── deployment-methods.md             # Basic / ConfigMgr / GPO / Ansible
    │       └── at-scale-onboarding.md
    ├── arc-server-troubleshoot/
    │   └── references/
    │       ├── agent-status.md                   # Connected / Disconnected / Expired / Error
    │       └── common-issues.md
    └── arc-server-manage/
        └── references/
            ├── agent-upgrade.md
            ├── extended-security-updates.md      # ESU
            ├── hotpatch.md
            └── pay-as-you-go.md                  # PayGo Windows Server
```

Content lifted from internal portal-team Arc skill source and adapted to this repo's conventions. All references point back to canonical `Microsoft.HybridCompute/{machines,licenses,gateways,privateLinkScopes}` ARM types and public learn.microsoft.com docs.

## Disambiguation from `azure-compute`

The `SKILL.md` description (908 chars, within the 1024 limit) is engineered to fire on Arc-specific signals (`on-prem`, `hybrid`, `azcmagent`, `Connected Machine`, `Arc-enabled`, ESU, Hotpatch, PayGo) and **not** on generic "create a VM in Azure" prompts. The integration suite hard-asserts `azure-compute` is **not** also invoked for an unambiguous on-prem prompt.

## Tests

`tests/azure-arc-servers/`:
- **`triggers.test.ts`** — 52 cases across 9 buckets: onboard / at-scale / connectivity / troubleshoot / manage / ESU / Hotpatch / PayGo / negatives / edge cases
- **`integration.test.ts`** — 4 routing assertions:
  1. Single onboard prompt → arc-servers
  2. On-prem prompt → arc-servers **and not** azure-compute
  3. "Disconnected" status → troubleshoot workflow
  4. "extended security updates" → manage workflow
- Snapshot file committed
- Registered in `tests/skills.json` (skills array + 12pm integration schedule slot)

## Char budget

Bumped `DEFAULT_SKILL_CHAR_BUDGET` in `scripts/src/copilot-cli-char-budget.ts` from `20000` → `22000`. Repo was already at 19815/20000 before this PR; the in-file comment explicitly says the constant is a soft cap meant to be bumped as new skills land. New total: **20875 / 22000** (≈1100 chars of headroom for the next contribution).

## Validation (all green locally)

| Check | Result |
|---|---|
| `npm run build` | ✅ stamps `azure-arc-servers 1.0.0` |
| `scripts: npm run frontmatter -- ../output/skills/azure-arc-servers/SKILL.md` | ✅ |
| `scripts: npm run references -- --skills-dir ../output/skills` | ✅ 28 / 28 |
| `scripts: npm run checkCopilotCliCharBudget` | ✅ 20875 / 22000 |
| `scripts: npm run checkPluginVersionPr` | ✅ |
| `scripts: npm test` | ✅ 463 / 463 |
| `tests: npm run typecheck && npm run lint` | ✅ |
| `tests: SKIP_INTEGRATION_TESTS=true npm test -- --testPathPatterns=azure-arc-servers` | ✅ 52 / 52 (+4 integration appropriately skipped) |

## Known soft warnings (non-blocking in CI)

- `npm run tokens check` flags `SKILL.md` (1394 tok) and `arc-server-onboard.md` (>2000 tok) above the *soft* default. CI only reports this; several existing skills (azure-compute, microsoft-foundry, entra-*, azure-upgrade) sit similarly over their soft caps. Happy to split onboarding further if a reviewer wants strict compliance.

Co-authored-by: Copilot <223556219+Copilot@users.noreply.github.com>

### CODE REVIEW

File: tests/azure-arc-servers/integration.test.ts [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+
+const SKILL_NAME = "azure-arc-servers";
+const ONBOARD_WORKFLOW_PATH =
+  /workflows\/arc-server-onboard\/arc-server-onboard\.md/i;
+const TROUBLESHOOT_WORKFLOW_PATH =
+  /workflows\/arc-server-troubleshoot\/arc-server-troubleshoot\.md/i;
+const MANAGE_WORKFLOW_PATH =
+  /workflows\/arc-server-manage\/arc-server-manage\.md/i;
+const RUNS_PER_PROMPT = 5;
+const invocationRateThreshold = 0.8;
```
[the remark this answers is not in the document]
Comment (deankroker) [reply tests/azure-arc-servers/integration.test.ts]: Addressed in edfa89c:
- Renamed `invocationRateThreshold` → `INVOCATION_RATE_THRESHOLD` ✅
- Kept the trailing underscore in `${SKILL_NAME}_ - Integration Tests` — it's required by the custom ESLint rule `integration-test-name` ([`tests/eslint-rules/integration-test-name.mjs`](https://github.com/microsoft/GitHub-Copilot-for-Azure/blob/main/tests/eslint-rules/integration-test-name.mjs)), which enforces the pattern `/^[a-z0-9-]+_[a-z0-9-]* - Integration Tests$/`. Every other skill in the repo uses this form. Removing it fails `npm run lint`.

File: tests/azure-arc-servers/integration.test.ts [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+const skipTests = shouldSkipIntegrationTests();
+const skipReason = getIntegrationSkipReason();
+
+if (skipTests && skipReason) {
+  console.log(`⏭️  Skipping integration tests: ${skipReason}`);
+}
+
+const describeIntegration = skipTests ? describe.skip : describe;
+
+describeIntegration(`${SKILL_NAME}_ - Integration Tests`, () => {
```
[the remark this answers is not in the document]
Comment (deankroker) [reply tests/azure-arc-servers/integration.test.ts:46]: See sibling thread — `INVOCATION_RATE_THRESHOLD` renamed; trailing underscore kept (required by ESLint rule `integration-test-name`).

File: tests/azure-arc-servers/triggers.test.ts [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+  describe("Should NOT Trigger - unrelated topics", () => {
+    const shouldNotTriggerPrompts: string[] = [
+      "What is the weather today?",
+      "Help me write a poem",
+      "Explain quantum computing",
+      "Help me with AWS Systems Manager",
+      "Configure my PostgreSQL database",
+      "How do I write a Python web scraper?",
+      "Set up a Kubernetes cluster with Helm",
+      "Create a serverless function on AWS",
```
[the remark this answers is not in the document]
Comment (deankroker) [reply tests/azure-arc-servers/triggers.test.ts:172]: Partially addressed in edfa89c.

Investigation: `TriggerMatcher` is a keyword pre-filter, not a router. It extracts keywords directly from SKILL.md, and arc-servers' SKILL.md inevitably contributes generic tokens — `azure`, `machine`, `server`, `windows`, `deploy`, `script` — that any Azure-VM prompt will hit. Three of the four prompts you suggested (`Deploy a new Windows VM in Azure ...`, `What VM size should I pick ...`, `Add a data disk to my Azure VM`) match ≥2 of these keywords and would assert as flaky-by-design.

Resolution: added a `Should NOT Trigger - unambiguous azure-compute` block with the 2 prompts that have no Arc-keyword overlap (`Create an Azure VM running Ubuntu in East US`, `Provision an Azure VM scale set for my web tier`), plus an inline comment documenting that real Arc-vs-Compute disambiguation is enforced by the LLM router and tested in `integration.test.ts` (the `'on-prem server to Azure' invokes arc-servers, NOT azure-compute` case asserts `wrongSkillInvocations === 0`). That's the right layer for this guarantee.

**Comment by ryperl** [comment]
Hey Dean -- looks like we need a CODEOWNERS entry for the new skill. Every existing skill has a line in `.github/CODEOWNERS` following this pattern:

```
/plugin/skills/{skill-name}/ @domain-owner(s) @RickWinter
```

Can you add one for `azure-arc-servers`? Here's two to start:

```
/plugin/skills/azure-arc-servers/ @deankroker @ryperl_msft @rebro2-msft @RickWinter
```

I'll get you Jackie's alias separately.

File: plugin/skills/azure-arc-servers/workflows/arc-server-onboard/references/prerequisites.md [hunk 4]
```diff
@@ -0,0 +1,120 @@
...
+- Local administrator / root to install the agent.
+- ~500 MB free disk under `%ProgramData%\AzureConnectedMachineAgent` or `/var/opt/azcmagent`.
+- TLS 1.2 enabled (most modern OSes; explicit on older Windows Server).
+
+### Conflicts to check before installing
+
+- The **Azure VM agent** (`Microsoft.Azure.HybridCompute` extension on
+  an Azure VM) is mutually exclusive with `azcmagent` on the same OS
+  instance. If the machine **is** an Azure VM, the user should not be
+  onboarding it via Arc.
```
[the remark this answers is not in the document]
Comment (deankroker) [reply plugin/skills/azure-arc-servers/workflows/arc-server-onboard/references/prerequisites.md]: Good catch — fixed in `16ad8568`. `Microsoft.Azure.HybridCompute` was wrong; the conflict note now talks about `Microsoft.Compute/virtualMachines` plus the Azure VM Guest Agent, and calls out that running the Connected Machine agent on an existing Azure VM isn't a supported configuration.

File: plugin/skills/azure-arc-servers/workflows/arc-server-manage/references/agent-upgrade.md [hunk 5]
```diff
@@ -0,0 +1,137 @@
...
+### CLI
+
+```bash
+# Read current setting (PATCH the desired state)
+az resource patch \
+    --resource-type "Microsoft.HybridCompute/machines" \
+    --resource-group <rg> \
+    --name <machine-name> \
+    --api-version 2025-02-19-preview \
+    --properties '{"agentUpgrade":{"enableAutomaticUpgrade":true}}'
```
[the remark this answers is not in the document]
Comment (deankroker) [reply plugin/skills/azure-arc-servers/workflows/arc-server-manage/references/agent-upgrade.md]: Fixed in `16ad8568` — the comment now reads "Enable automatic agent upgrade (PATCH sets the desired state)", and I added a separate `az resource show ... --query "properties.agentUpgrade.enableAutomaticUpgrade"` example to read the current value.

File: plugin/skills/azure-arc-servers/SKILL.md [hunk 6]
```diff
@@ -0,0 +1,87 @@
+---
+name: azure-arc-servers
+description: "Azure Arc-enabled servers router. WHEN: Arc Server, Arc-enabled server, Connected Machine agent, azcmagent, hybrid server, on-prem server, onboard to Arc, project a server into Azure, connect machine to Azure, install Connected Machine agent, generate onboarding script, at-scale onboarding, Group Policy onboarding, Configuration Manager onboarding, Ansible onboarding, Service Principal for Arc, Arc Private Link Scope, Arc Gateway, proxy for Arc, agent connectivity, agent status, Connected / Disconnected / Expired / Error, agent upgrade, automatic upgrade, manual upgrade, Extended Security Updates (ESU) for Arc, Hotpatch on Arc, Pay-as-you-go Windows Server, Software Assurance benefits, Microsoft.HybridCompute, hybridcompute/machines. PREFER OVER azure-compute when the user is talking about non-Azure (on-prem, other-cloud, edge) servers being projected into Azure - not about creating an Azure VM."
```
[the remark this answers is not in the document]
Comment (deankroker) [reply plugin/skills/azure-arc-servers/SKILL.md:3]: Leaving the description as-is. Naming the sibling skill is an established convention in this repo — `azure-hosted-copilot-sdk` names `azure-prepare`/`azure-deploy`/`microsoft-foundry`, and `azure-compute` names `mcp__azure__get_azure_bestpractices`. The `azure-compute` token only matches if a prompt literally contains "azure-compute" (effectively never), and the disambiguation signal to the router outweighs that marginal false-positive risk. Resolving.

File: tests/azure-arc-servers/integration.test.ts [hunk 7]
```diff
@@ -0,0 +1,153 @@
...
```
[the remark this answers is not in the document]
Comment (deankroker) [reply tests/azure-arc-servers/integration.test.ts:46]: The trailing underscore is required by our custom ESLint rule `integration-test-name` (pattern `^[a-z0-9-]+_[a-z0-9-]* - Integration Tests$`). Removing it fails `npm run lint`, and every skill in the repo uses this form. Resolving as working-as-intended.

**Comment by deankroker** [comment]
Added in `c435aab6`:

```
/plugin/skills/azure-arc-servers/ @deankroker @ryperl_msft @rebro2-msft @jackier @RickWinter
```

Thanks @ryperl — folded in Jackie (`@jackier`) as well.

File: tests/azure-arc-servers/triggers.test.ts [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+      expect(triggerMatcher.getKeywords()).toMatchSnapshot();
+    });
+
+    test("skill description triggers match snapshot", () => {
+      expect({
+        name: skill.metadata.name,
+        description: skill.metadata.description,
+        extractedKeywords: triggerMatcher.getKeywords(),
+      }).toMatchSnapshot();
+    });
```
Comment (jongio) [inline tests/azure-arc-servers/triggers.test.ts:225]: The near-neighbor section only has 2 azure-compute prompts. Since disambiguation from azure-compute is a primary design goal (called out in the PR description, the SKILL.md blockquote, and the integration tests), consider expanding this. Cases like:

- "Deploy a Windows Server VM in Azure"
- "What VM size should I use for my web app?"
- "Resize my Azure virtual machine"
- "Set up autoscale for my VMSS"

These would strengthen confidence that the keyword extractor doesn't false-positive on VM-adjacent language when Arc-specific signals are absent.

File: plugin/skills/azure-arc-servers/SKILL.md (hunk 6 above)
Comment (jongio) [inline plugin/skills/azure-arc-servers/SKILL.md:3]: This description is 908 chars on a single line. The `tests/AGENTS.md` frontmatter rule says descriptions over 200 chars should use folded YAML (`>-`) for readability. Existing skills like azure-compute also use single-line format, so this isn't blocking, but `>-` would make future edits easier:

```suggestion
description: >-
  Azure Arc-enabled servers router. WHEN: Arc Server, Arc-enabled server,
  Connected Machine agent, azcmagent, hybrid server, on-prem server,
  onboard to Arc, project a server into Azure, connect machine to Azure,
  install Connected Machine agent, generate onboarding script, at-scale
  onboarding, Group Policy onboarding, Configuration Manager onboarding,
  Ansible onboarding, Service Principal for Arc, Arc Private Link Scope,
  Arc Gateway, proxy for Arc, agent connectivity, agent status,
  Connected / Disconnected / Expired / Error, agent upgrade, automatic
  upgrade, manual upgrade, Extended Security Updates (ESU) for Arc,
  Hotpatch on Arc, Pay-as-you-go Windows Server, Software Assurance
  benefits, Microsoft.HybridCompute, hybridcompute/machines. PREFER OVER
  azure-compute when the user is talking about non-Azure (on-prem,
  other-cloud, edge) servers being projected into Azure - not about
  creating an Azure VM.
```

File: scripts/src/copilot-cli-char-budget.ts [hunk 9]
```diff
@@ -8,9 +8,9 @@ import { escapeXml } from "./shared/string-helpers.js";
 
 const SKILL_CHAR_BUDGET_ENV = "SKILL_CHAR_BUDGET";
 // Note: Copilot CLI's default skill char budget is 15000,
-// we set our budget to 20000 as a soft cap to not block skill contributions
+// we set our budget to 22000 as a soft cap to not block skill contributions
 // and remind us when the consumption grows too fast.
-const DEFAULT_SKILL_CHAR_BUDGET = 20000;
+const DEFAULT_SKILL_CHAR_BUDGET = 22000;
```
Comment (jongio) [inline scripts/src/copilot-cli-char-budget.ts:13]: The bump from 20000 to 22000 (+10%) is reasonable given the PR description says the repo was at 19815 before this skill. Worth confirming the new total (20875/22000) still holds after rebasing on main, since other skill PRs may have merged and changed the baseline.

**Review by jongio — COMMENTED** [review]
The skill structure is solid: clear routing, thorough workflows, and good disambiguation from azure-compute.

Two things need resolution before merge:

1. **CI failures**: ESLint and eval checks are failing on the latest commit. These need to pass.
2. **Merge conflicts**: The PR currently conflicts with main and needs a rebase.

A few inline comments below on the tests and SKILL.md formatting.
